libminifi/test/unit/MinifiConcurrentQueueTests.cpp

```diff
@@ -29,132 +29,290 @@
-      results.push_back(s);
-    }
-  });
+  std::thread getReaddingDequeueConsumerThread(utils::ConditionConcurrentQueue<std::string>& queue, std::vector<std::string>& results) {
+    return std::thread([&queue, &results] {
+      std::string s;
+      while (queue.dequeueWait(s)) {
+        if (!std::count(results.begin(), results.end(), s)) {
+          results.push_back(s);
+        }
+        // The consumer is busy enqueing so noone is waiting for this ;(
+        queue.enqueue(std::move(s));
+      }
+    });
+  }
 
-  producer.join();
+  std::thread getDequeueWaitForConsumerThread(utils::ConditionConcurrentQueue<std::string>& queue, std::vector<std::string>& results) {
+    return std::thread([&queue, &results] {
+      const std::size_t max_read_attempts = 6;
+      std::size_t attemt_num = 0;
+      while (results.size() < 3 && attemt_num < max_read_attempts) {
+        ++attemt_num;
+        std::string s;
+        if (queue.dequeueWaitFor(s, std::chrono::milliseconds(3))) {
+          results.push_back(s);
+        }
+      }
+    });
+  }
+
+  std::thread getConsumeWaitForConsumerThread(utils::ConditionConcurrentQueue<std::string>& queue, std::vector<std::string>& results) {
+    return std::thread([&queue, &results]() {
+      const std::size_t max_read_attempts = 6;
+      std::size_t attemt_num = 0;
+      while (results.size() < 3 && attemt_num < max_read_attempts) {
+        ++attemt_num;
+        std::string s;
+        queue.consumeWaitFor([&results] (const std::string& s) { results.push_back(s); }, std::chrono::milliseconds(3));
+      }
+    });
+  }
 
-  std::this_thread::sleep_for(std::chrono::milliseconds(10));
+}  // namespace MinifiConcurrentQueueTestProducersConsumers
 
-  queue.stop();
+TEST_CASE("TestConcurrentQueue", "[TestConcurrentQueue]") {
+  using namespace MinifiConcurrentQueueTestProducersConsumers;
 
-  consumer.join();
+  utils::ConcurrentQueue<std::string> queue;
+  std::vector<std::string> results;
 
-  REQUIRE(utils::StringUtils::join("-", results) == "ba-dum-tss");
-}
+  SECTION("empty queue") {
+    SECTION("default initialized queue is empty") {
+      REQUIRE(queue.empty());
+    }
 
+    SECTION("trying to update based on empty queue preserves original data") {
+      std::string s { "Unchanged" };
 
-/* In this testcase the consumer thread puts back all items to the queue to consume again
- * Even in this case the ones inserted later by the producer  should be consumed */
-TEST_CASE("TestConqurrentQueue::testQueueWithReAdd", "[TestQueueWithReAdd]") {
-  utils::ConcurrentQueue<std::string> queue;
-  std::set<std::string> results;
-
-  std::thread producer([&queue]() {
-    queue.enqueue("ba");
-    std::this_thread::sleep_for(std::chrono::milliseconds(3));
-    queue.enqueue("dum");
-    std::this_thread::sleep_for(std::chrono::milliseconds(3));
-    queue.enqueue("tss");
-  });
+      SECTION("tryDequeue on empty queue returns false") {
+        REQUIRE(false == queue.tryDequeue(s));
```

Comment:
clion inspection: "Expression can be simplified to `! queue.tryDequeue(s)`"

Reply:
Sure, I'll replace this, I am just used to writing tests in `expected, actual` notation.

Reply:
Updated as requested.

Reply:
Many of my code review comments point out things that can be done in a shorter/simpler way if there are no reasons for doing otherwise. Your preference for the "expected, actual" notation is a good enough reason for me to ignore this inspection. Feel free to question my suggestions with valid points.